### PR TITLE
Add ListView text overflow

### PR DESCRIPTION
**Pull request by tiffany-sy-lin**
GH-3052

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

1. Set overflowMode="wrap" in ListView. 
2. Create a long title/de
3. scription/content. 
4. Set a width for the ListView.
5. See that the text wraps the next line
6. Do the same where overflowMode="truncation" or when no overflowMode is set
7. See that the text is cut off and an ellipsis is applied

## 🧢 Your Project:
Quarry (we currently have some custom css to mimic this behavior in Inventory)

### CODE REVIEW

File: packages/@react-spectrum/list/src/ListView.tsx [hunk 1]
```diff
@@ -201,7 +205,7 @@ function ListView<T extends object>(props: ListViewProps<T>, ref: DOMRef<HTMLDiv
   }
 
   return (
-    <ListViewContext.Provider value={{state, keyboardDelegate, dragState, onAction, isListDraggable, layout}}>
+    <ListViewContext.Provider value={{state, keyboardDelegate, dragState, onAction, isListDraggable, layout, overflowMode}}>
```
Comment (snowystinger) [inline packages/@react-spectrum/list/src/ListView.tsx:208]: could we do the overflow mode in one place like Table does? https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/table/src/TableView.tsx#L302
or did that not work?

File: packages/@react-spectrum/list/test/ListView.test.js [hunk 2]
```diff
@@ -1197,4 +1197,32 @@ describe('ListView', function () {
...
+      <ListView aria-label="list" width="250px" overflowMode="wrap">
+        <Item textValue="row 1">row 1 with a very very very very very long title</Item>
+        <Item textValue="row 2">row 2</Item>
+        <Item textValue="row 3">row 3</Item>
+      </ListView>
+    );
+
+    const items = getAllByRole('gridcell');
+    items.forEach((item) => {
+      expect(item.children[0].children[0]).toHaveClass('react-spectrum-ListViewItem--wrap');
```
Comment (snowystinger) [inline packages/@react-spectrum/list/test/ListView.test.js:1211]: This test isn't needed, we'll get a chromatic story to cover this instead

Comment (snowystinger) [reply packages/@react-spectrum/list/test/ListView.test.js:1211]: same thing with test below

File: packages/@react-spectrum/list/stories/ListView.stories.tsx [hunk 3]
```diff
@@ -242,8 +242,21 @@ storiesOf('ListView', module)
...
+      <Item textValue="row 3">row 3</Item>
+    </ListView>
+  ))
+  .add('overflowMode="wrap"', () => (
+    <ListView width="250px" overflowMode="wrap">
+      <Item textValue="row 1">row 1 with a very very very very very long title</Item>
+      <Item textValue="row 2">row 2</Item>
+      <Item textValue="row 3">row 3</Item>
+    </ListView>
+  ));
```
Comment (LFDanLu) [inline packages/@react-spectrum/list/stories/ListView.stories.tsx:259]: mind adding long description, link, etc in these stories as well? Will help us test all the cases

File: packages/@react-spectrum/list/src/ListView.tsx (hunk 1 above)
Comment (tiffany-sy-lin) [reply packages/@react-spectrum/list/src/ListView.tsx:208]: I tried that at first, but the grid and draggable stuff got between the 'wrap' class and the item class. I also saw a conditional class being assigned so was worried that would also affect it. And I didn't want to have a super long css class identifier.

**Review by snowystinger — APPROVED** [review]
lgtm, tried new stories in chrome/safari/ff

**Review by ktabors — DISMISSED** [review]
For some reason I thought we'd be adding the chromatic stories now and not later. These stories and functionality are so simple I'd rather they exist there than adding to the stories we have in storybook.

I've been starting at selection and DnD stories too much, I want to see selection checkboxes and buttons in these wrapped or truncated items too. I'm also thinking of this because Rob shared a story with narrow ListView with buttons and it didn't look great.

The change is great, testing on desktop Chrome and iPhone.

**Comment by devongovett** [comment]
When `overflowMode="truncate"`, we know all the rows will be the same height. This provides an opportunity for a performance optimization: since the rows are equal height, we can avoid measuring them when they come into view while scrolling. This is controlled by the `ListLayout`, which returns either a fixed height or an estimated height, which is later replaced with the real height after scrolling an item into view.

https://github.com/adobe/react-spectrum/blob/cf861672e2bc05d52fad1a8b39775327b3816250/packages/%40react-spectrum/list/src/ListView.tsx#L73

The above can be changed to `rowHeight` rather than `estimatedRowHeight` when `overflowMode="truncate"`, similar to how we do it in TableView:

https://github.com/adobe/react-spectrum/blob/cf861672e2bc05d52fad1a8b39775327b3816250/packages/%40react-spectrum/table/src/TableView.tsx#L127-L133

**Comment by ktabors** [comment]
If truncate is default, can we have the long strings as part of another story so we only add one story which is the wrap with long strings?

File: packages/@react-spectrum/list/src/ListView.tsx [hunk 4]
```diff
@@ -63,13 +63,13 @@ const ROW_HEIGHTS = {
...
 
-function useListLayout<T>(state: ListState<T>, density: ListViewProps<T>['density']) {
+function useListLayout<T>(state: ListState<T>, density: ListViewProps<T>['density'], overflowMode: ListViewProps<T>['overflowMode']) {
   let {scale} = useProvider();
   let collator = useCollator({usage: 'search', sensitivity: 'base'});
   let isEmpty = state.collection.size === 0;
   let layout = useMemo(() =>
     new ListLayout<T>({
-      estimatedRowHeight: ROW_HEIGHTS[density][scale],
+      estimatedRowHeight: overflowMode === 'wrap' ? null : ROW_HEIGHTS[density][scale],
```
Comment (snowystinger) [inline packages/@react-spectrum/list/src/ListView.tsx:72]: from devon's comment, ListLayout accepts both estimatedRowHeight as well as rowHeight
https://github.com/adobe/react-spectrum/blob/main/packages/%40react-stately/layout/src/ListLayout.ts#L20
when in overflowMode === 'wrap', we should use estimatedRowHeight as we were before
when in overflowMode === 'truncate', we should use rowHeight with the same value

**Review by LFDanLu — APPROVED** [review]
Pushed a minor update for the overflowMode type definition, looks good to me.
The truncate storybook example has revealed a iffy case but that can be handled in a follow up IMO:
image

**Comment by tiffany-sy-lin** [comment]
> Pushed a minor update for the overflowMode type definition, looks good to me. The truncate storybook example has revealed a iffy case but that can be handled in a follow up IMO: image

i think it should be handled now since this the entire point of the PR is to implement overflow

**Review by LFDanLu — APPROVED** [review]
LGTM

**Pull request merged**